### DISCUSSION TITLE
iOS编译失败

### DISCUSSION THREAD

**Discussion by lay-225** [alibaba/xquic#35] [Q&A] [answered]
应交流群大佬要求 在此处贴出编译失败问题

系统版本：
    Apple M1 Pro macos 12.1。

gcc版本：
    Configured with: --prefix=/Library/Developer/CommandLineTools/usr --with-gxx-include-dir=/Library/Developer/CommandLineTools/SDKs/MacOSX12.1.sdk/usr/include/c++/4.2.1
    Apple clang version 13.0.0 (clang-1300.0.27.3)
    Target: arm64-apple-darwin21.2.0
    Thread model: posix
    InstalledDir: /Library/Developer/CommandLineTools/usr/bin、

cmake版本：
    cmake version 3.22.1
    CMake suite maintained and supported by Kitware (kitware.com/cmake).

make版本：
    GNU Make 3.81
    Copyright (C) 2006  Free Software Foundation, Inc.
    This is free software; see the source for copying conditions.
    There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A
    PARTICULAR PURPOSE.
    This program built for i386-apple-darwin11.3.0、

IOS_CMAKE_TOOLCHAIN：
    https://github.com/leetal/ios-cmake

xqc_build.sh 脚本改动：
        generate_plat_spec() {
        plat_spec=
        if [ x"$platform" == xios ] ; then
            plat_spec="-DARCHS=$1"
            if [ x"$1" == xarmv7 ] ; then
                plat_spec="$plat_spec -DPLATFORM=OS"
            elif [ x"$1" == xarm64 ] ; then
                plat_spec="$plat_spec -DPLATFORM=OS64"
            elif [ x"$1" == xx86_64 ] ; then
                plat_spec="$plat_spec -DPLATFORM=SIMULATOR64"
            elif [ x"$1" == xi386 ] ; then
                plat_spec="$plat_spec -DPLATFORM=SIMULATOR"
            fi
        else
            plat_spec="-DANDROID_ABI=$1"
        fi
        echo $plat_spec
    }

build命令： sh xqc_build.sh ios  build-ios build-ios-target ./third_party/boringssl

报错信息如下：

        create build directory(build-ios) suc
        create artifact directory(build-ios-target) suc
        compiling xquic on armv7 arch
        CMake Deprecation Warning at CMakeLists.txt:3 (cmake_minimum_required):
        Compatibility with CMake < 2.8.12 will be removed from a future version of
        CMake.

        Update the VERSION argument <min> value or use a ...<max> suffix to tell
        CMake that the project does not need compatibility with older versions.

        -- [DEFAULTS] Enabling ARC support by default. ENABLE_ARC not provided!
        -- [DEFAULTS] Hiding symbols visibility by default. ENABLE_VISIBILITY not provided!
        -- [DEFAULTS] Using NON-strict compiler checks by default. ENABLE_STRICT_TRY_COMPILE not provided!
        CMake Error at ios.toolchain.cmake:550 (get_filename_component):
        get_filename_component called with incorrect number of arguments
        Call Stack (most recent call first):
        /opt/homebrew/Cellar/cmake/3.22.1/share/cmake/Modules/CMakeDetermineSystem.cmake:124 (include)
        CMakeLists.txt:4 (project)

        CMake Error: CMake was unable to find a build program corresponding to "Unix Makefiles".  CMAKE_MAKE_PROGRAM is not set.  You probably need to select a different build tool.
        CMake Error: CMAKE_C_COMPILER not set, after EnableLanguage
        CMake Error: CMAKE_CXX_COMPILER not set, after EnableLanguage
        -- Configuring incomplete, errors occurred!
        make: *** No targets specified and no makefile found.  Stop.

**Comment by ruiqizhou** [comment, marked as answer]
It seems to be an xcode environment issue. Did you install xcode? 

try `/cmake/ios.toolchain.cmake`

**Reply by lay-225** [reply]
没有安装xcode ide  但是安装了  xcode command line tools。  
我执行 xcode-select  --install  提示我已经安装
$ xcode-select  --install
xcode-select: error: command line tools are already installed, use "Software Update" to install updates

我没理解  /cmake/ios.toolchain.cmake  是什么操作

**Reply by ruiqizhou** [reply]
export IOS_CMAKE_TOOLCHAIN="xx/xquic/cmake/ios.toolchain.cmake"

![image](https://user-images.githubusercontent.com/95687026/149097274-fa609340-648c-476c-bc2d-9f4d26255655.png)

**Reply by lay-225** [reply]
https://github.com/leetal/ios-cmake  跟这个一样

**Reply by ruiqizhou** [reply]
> https://github.com/leetal/ios-cmake 跟这个一样

The two files are different, I think you should try `/cmake/ios.toolchain.cmake` first

**Reply by lay-225** [reply]
我更换了 ，报错行数从550行  变成了404行  错误信息一模一样

**Comment by ruiqizhou** [comment]
This discussion will also help you: #21.

**Reply by lay-225** [reply]
configures="-DSSL_TYPE=${ssl_type}
 47                 -DSSL_PATH=${ssl_path}
 48                 -DSSL_LIB_PATH=${ssl_lib_path}
 49                 -DDEPLOYMENT_TARGET=10.0
 50                 -DCMAKE_BUILD_TYPE=Minsizerel
 51                 -DXQC_ENABLE_TESTING=OFF
 52                 -DXQC_BUILD_SAMPLE=OFF
 53                 -DGCOV=OFF
 54                 -DCMAKE_TOOLCHAIN_FILE=${IOS_CMAKE_TOOLCHAIN}
 55                 -DENABLE_BITCODE=0
 56                 -DXQC_NO_SHARED=1"
 57

不行 没有解决  我的脚本中早已经删除了

**Comment by ruiqizhou** [comment]
And yesterday you said in the DingTalk group that `xcode 1.5 no support` , so we think there is something wrong with your xcode environment. See https://blog.csdn.net/sun007700/article/details/97943542

**Reply by lay-225** [reply]
出现   xcode 1.5 no support  那个是因为我加了一个错误的编译参数 -G Xcode(加这个的原因  是一开始就报了上述错误<CMake Error at ios.toolchain.cmake:550 (get_filename_component):>，病急乱投医了) 。之后群里成功的说是不需要这个参数的。 xcode command line tools 我也已经更新到了最新

**Reply by ruiqizhou** [reply]
did u try `export IOS_CMAKE_TOOLCHAIN="xx/xquic/cmake/ios.toolchain.cmake"` ?

**Reply by lay-225** [reply]
![image](https://user-images.githubusercontent.com/42629375/149124638-e1f2de3d-943c-4b46-8e81-20360e722a44.png)

**Reply by ruiqizhou** [reply]
@lurker-Chen

**Reply by lay-225** [reply]
我修改了环境变量值 现在是/Users/leizihui/Documents/xquic/cmake/ios.toolchain.cmake  仅仅是报错的行数变了

![image](https://user-images.githubusercontent.com/42629375/149126673-10ec26a9-eeec-4d46-a74e-75197d4c9f7b.png)

**Reply by lurker-Chen** [reply]
it seems that your CMAKE_OSX_SYSROOT automatically determined from PLATFORM and xcodebuild is incorrect. specify it manually in xquic_build.sh

**Reply by lay-225** [reply]
我不太清楚  CMAKE_OSX_SYSROOT  是什么目录。

**Reply by lurker-Chen** [reply]
check ios.toolchain.cmake
![image](https://user-images.githubusercontent.com/97525334/149268223-f98d6d14-fdfe-4131-932f-d477fc4ada32.png)

**Comment by lay-225** [comment]
完整安装xcode app才行，只安装xcode command line tools不行。 感谢大佬！

**Reply by lurker-Chen** [reply]
The Xcode app includes iOS SDKs. Seems you only have xcode command line tools before, but no any iOS SDK?

**Reply by lay-225** [reply]
看起来  xcode command line tools  没有携带iOS  sdk。 编译xquic工程需要装xcode ide。。
